### PR TITLE
MBT Java 8 Docker Image

### PR DESCRIPTION
**Pull request by young-yang03**
## Description

Add Java 8 Docker Image build and publish.

Fix npm packages global installation #1026 

Add a link to the design (if applicable).

### Checklist
- [x] Code compiles correctly
- [x] Relevant tests were added (unit / contract / integration)
- [x] Relevant logs were added
- [x] Formatting and linting run locally successfully
- [x] All tests pass
- [x] UA review
- [x] Design is documented
- [x] Extended the README / documentation, if necessary
- [x] Open source is approved

### CODE REVIEW

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:26]: The GNU/Linux file hierarchy standard (aka FHS) does not include all that paths. That's why Dockerfile usually creates symlinks to FHS paths instead of playing with $PATH env variable to avoid introducing security issues. 

Defining the installation path for Go, Node.js, Maven is fine (but not strictly needed) and is out of that PR scope and shall be done in another PR. And given the extra commands added to cope with them that introduce permission issues, not thought enough to be part of that PR. Please revert. That PR should only focus on integrating the java 8 and nothing more. The original PR was doing exactly that, please just take it as it was, do any minor fixes to it found during our tests and merge it.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:85]: You are mixing apples and tomatoes: GOOS variable is in the Go namespace, not node.js one.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:91]: Security issue: no user need to access the node.js installation path.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:90]: cf. previous comment.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:154]: The original shell code was working perfectly without introducing any permission security issues, unneeded symlinks, unneeded and unused for building env variables. Please revert. 

cf. previous comment.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:201]: No, you can't download the hash from the same source for obvious security reasons: if the source is compromised, both binaries and hash will be changed.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:245]: --prefix is unneeded.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:191]: cf. previous comment.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:205]: Still mixing apples and tomatoes in the env variable namespace ...

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:204]: Unneeded

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:24]: No needed, it duplicates GO_HOME purpose

**Review by jerome-benoit — DISMISSED** [review]

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:24]: GOPATH env already exist in the past (https://github.com/SAP/cloud-mta-build-tool/commit/35393669a1319333777bf26c864976307a496f0b#diff-3632a131a98f4a6db44b6238249146e491fa94be0b683a407b11498c3bffc11c)

when we merge all docker files into Dockerfile_mbtci_template, GOPATH is deleted, now I just revert it.

As to GO_HOME env, I will revert it, and when install go, use GO_HOME as a local var

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:26]: I will revert the GO_HOME env, and when install go, use GO_HOME as a local var

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:85]: yes

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:205]: yes

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:90]: I change nodejs install target path, from /usr/local/bin to /usr/local/nodejs. Same as go, from /usr/local/bin to /usr/local/go.

Making node and go installed different path is clear, and before we merge all docker file into one docker file Dockerfile_mbtci_template, the implementaion in Dockerfile_mbtci_java11, java8 and java 17, different modules are installed in different folder.

Install path is invisible to docker image users, why changing the install path may cause "backwards compatibility" ?

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:154]: In the past implementation before we merge all Docker files into Dockerfile_mbtci_template, $GOPATH path is set 777 permission, I am afraid some user may reply on the write permission, so I rever that code.

You can ref : https://github.com/SAP/cloud-mta-build-tool/commit/35393669a1319333777bf26c864976307a496f0b#diff-3632a131a98f4a6db44b6238249146e491fa94be0b683a407b11498c3bffc11c

And, for the new added GO_HOME env, you are right, I will change it to a local var and will not export it as a ENV

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:201]: In current implementation, we hard code the compare of sapjvm-${SAPMACHINE_VERSION}-linux-${ARCH}.zip's sha1 value with "583dedfeee0c119839a610ddc8c6768d0c044429".

My comment "# Notice to modify, download the sha1 value auto" means, in the future, we will get the sha1 value from link : "https://tools.hana.ondemand.com/additional/sapjvm-8.1.091-linux-x64.zip.sha1", not hard code the sha1 value "583dedfeee0c119839a610ddc8c6768d0c044429". 

I can delete my comment first, and update the code later

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:24]: It was not needed at that time also. JAVA_HOME is the only env variable needed (defined by the java standard). 

The goal is to keep thing simple and straightforward, so adding paths indirection in the image via symlinks or env vars only and only if strictly needed, the filesystem hierarchy standard is already fulfilled in the source archive. 
With all due respect, the original Dockerfile does not have anything in it that shall be considered as material for its rewrite. 

And you should consider not redoing PR already done and working as a maintainer. You just have to comment and wait for contributors to take comments into account and run the CI validation on PR, like for example the maintainer has done here: https://github.com/carlossg/docker-maven/pull/304. And then merge. 
The maintainer workload is already enough without adding to it unneeded PR redo extra work.

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:91]: chmod --recursive a+w "${NODE_HOME}" is added because of a customer issue https://github.com/SAP/cloud-mta-build-tool/issues/1018

In the case, customer exec "npm link @sap/cds-dk --local" in /home/mta folder, it will raise a access problem , such as :"permission denied, mkdir '/usr/local/lib/node_modules".

In my commit, I install nodejs in "/usr/local/nodejs", and to solve the customer's problem, I set the nodejs folder "a+w"  right

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:90]: The docker image must follow established standard expected by users to be followed like the FHS (https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard). 
You can create dedicated installation path for the archive installed but not in /usr/local. But it's unneeded and just add unneeded env vars, shell commands, etc. and must be done as part of another PR. 

For example `npm -g install` just follow the FHS and install binaries in /usr/local to keep things simple. Since all archives follow the FHS (with the expected permission) and are designed to be unarchived into /usr/local to avoid playing with PATH (and can lead to security issue), what is point of making things more complicated than it should at the cost of critical env var mangling such as PATH and insecure permissions change?

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:201]: You can't do it for obvious security reason:  if the source is compromised, both binaries and hash will be changed.

File: Dockerfile_mbtci_template
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:91]: 1. the issue was already solved by the use of `--no-same-owner ` and the correct fix is not chmod 777 but chown 
2. fixing bug unrelated to the PR purpose must be done in another PR
3. installation path change is not the purpose of the PR and must be done in another PR
4. chmod 777 on directory is not compliant with the SAP security standards 

And I've already explained at large in several comments why 3 is not necessary with in depth rationale.

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template:204]: No. if we do not chown to root, the unzipped sapjvm_8 folder's owner is not root, it is wired that the folder's owner is "kmem"

Comment (jerome-benoit) [reply Dockerfile_mbtci_template:204]: The clean fix is to use the `--no-same-owner` of bsdtar. 
 
And you just have to review, test and comment the original PR. Redoing a PR is not needed and add extra workload to your maintenance work. You have learn to rely on external contributions and make sure you can merge them safely after review by making the CI on them works. The maintainer duty is already too important as it as without add to it PR redo.

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: I revert the code which change owner for /usr/local/lib/node_modules path from PR #1023 in my commit https://github.com/SAP/cloud-mta-build-tool/pull/1023/commits/c449c37d3318ae82808dd931157c37c0315b41c5 , and this problem should be fixed in your PR #1026 

@rodibrin

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: Yes, I revert the code which install nodejs under path "/usr/local/nodejs", now nodejs is installed under "/usr/local"
https://github.com/SAP/cloud-mta-build-tool/pull/1023/commits/b3ea0b94387d71af5ac4e4952a133f0a78328619

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: I revert the code which change go install path to "/usr/local/go", now go is installed under "/usr/local"

https://github.com/SAP/cloud-mta-build-tool/pull/1023/commits/b3ea0b94387d71af5ac4e4952a133f0a78328619

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: Ok, get it.

Hard code "583dedfeee0c119839a610ddc8c6768d0c044429" can insure we download the correct binary.

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: OK, delete $MAVEN_HOME from path and revert "ln -s ${MAVEN_HOME}/bin/mvn /usr/bin/mvn"

https://github.com/SAP/cloud-mta-build-tool/pull/1023/commits/b3ea0b94387d71af5ac4e4952a133f0a78328619

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: fix it in https://github.com/SAP/cloud-mta-build-tool/commit/b3ea0b94387d71af5ac4e4952a133f0a78328619

File: Dockerfile_mbtci_template
Comment (young-yang03) [reply Dockerfile_mbtci_template]: yes, fix it in https://github.com/SAP/cloud-mta-build-tool/commit/b3ea0b94387d71af5ac4e4952a133f0a78328619

**Review by young-yang03 — COMMENTED** [review]
finish adding java8 docker image support and all code problem fix

**Pull request merged**
